### ISSUE TITLE
Heatmaps: how to draw heatmaps?

### ISSUE THREAD

**Issue by henuJackSon** [sstary/SSRS#62]
Hello
I came to the GitHub homepage of your laboratory through the following link. In the code of 'heatmap.py' of this link, some parameters are wrong, such as the parameter 'image_patches'. Do you have the correct code? Or can you provide code for displaying heat maps suitable for change detection?
Looking forward to your reply
Thank you
The link is as follows:
https://github.com/udrs/GCD/blob/main/heatmap.py

**Comment by sstary** [comment]
Hello, this link is a modified version by other researchers. 
You can copy the correct version in the SSRS/ MFNet/UNetFormer_MMSAM_heatmap.py path. This is actually a very simple way to draw it. As shown in the SSRS/MFNet /UNetFormer_MMSAM_heatmap.py file, all you need to do is record the feature map of the heatmap you are trying to draw in your model and pass it out to the test function. The SSRS/MFNet /test_heatmap.py shows the actual drawn code, but note that you need to change the batch size to 1, and you cannot call a declaration like torch.with no grad that does not preserve the computed gradient.
I suggest you modify the way you draw on your model accordingly, and bring up any other bugs if you have in future.

**Comment by henuJackSon** [comment]
Thank you very much for your timely answer.
I'm going to do the experiment right now, and I have one more question is do I need to use higher-dimensional (B,C,H,W) features to draw the heat map? For example, (4,512,H,W) still use the feature map after the decoder, for example, I want to classify 5, the feature map output from the decoder (4,5,H,W).
Thanks again.

**Comment by sstary** [comment]
You can draw feature maps in any dimension of the model. But the batch size must be 1. That is, your feature map must conform to [1, C, h, w], and c, h, w is not restricted.

**Comment by henuJackSon** [comment]
Hello, sorry to bother you again, I saved the decoder output feature map, shape is (5,416,416), and then run the error encountered in the process, the code has been modified. At present, I have tried to display the output result of the encoder (shape is (512,416,416)) for thermal map display, and the output result of the decoder (shape is (5,416,416)) for thermal map display, and the output result is all red. My code is as follows:
```python
import matplotlib.pyplot as plt
import torch.nn as nn
import torch.nn.functional as F
import torch.nn.init
import torch.autograd as autograd
import cv2
import numpy as np

# 加载特征图,此处获取并转换为特征图
# 从模型输出的特征图保存的是 "feature_map.npy"，形状保存为 (channels, height, width) 格式
feature = np.load("xxxx/outnpy/90.npy")  # 加载存储的特征图
feature = torch.tensor(feature, device='cuda')  # 将其转换为 Torch tensor，确保在正确设备上
print("第一次输出的feature的形状是",feature.shape)
pred_target2 = feature
pred2 = F.softmax(pred_target2, dim=0)
outs = pred2

x_comp = 100
y_comp = 100

feature = outs[:, :, :].requires_grad_(True)
pred = feature[1, x_comp, y_comp]  # 这里的 pred 是一个标量 Tensor
feature_grad = autograd.grad(pred, feature, allow_unused=True, retain_graph=True)[0]

grads = feature_grad  # 获取梯度
pooled_grads = torch.nn.functional.adaptive_avg_pool2d(grads, (1, 1))

# 克隆整个 feature，不丢失通道维度
cloned_feature = feature.clone()
# 确保 pooled_grads 的形状匹配
print("pooled_grads形状",pooled_grads.shape)
pooled_grads = pooled_grads.squeeze()  # 去掉无关的维度，变成 [2]
print("cloned_feature形状",cloned_feature.shape)

print("Before scaling:")
print(cloned_feature[0, 100, 100])  # 打印第一个通道的某个像素值
print(cloned_feature[1, 200, 200])  # 打印第二个通道的某个像素值

print("Pooled grads:")
print(pooled_grads[0])  # 第一个通道的加权因子
print(pooled_grads[1])  # 第二个通道的加权因子
print(pooled_grads[2])  # 第三个通道的加权因子
print(pooled_grads[3])  # 第四个通道的加权因子
print(pooled_grads[4])  # 第五个通道的加权因子

for i in range(cloned_feature.shape[0]):
    cloned_feature[i, ...] *= pooled_grads[i, ...]

print("After scaling:")
print(cloned_feature[0, 100, 100])  # 加权后的值
print(cloned_feature[1, 200, 200])  # 加权后的值

heatmap = feature.detach().cpu().numpy()
heatmap = np.mean(heatmap, axis=0)
heatmap1 = np.maximum(heatmap, 0)
heatmap1 /= np.max(heatmap1)
heatmap1 = cv2.resize(heatmap1, (256, 256))
# heatmap[heatmap < 0.7] = 0
heatmap1 = np.uint8(255 * heatmap1)
heatmap1 = cv2.applyColorMap(heatmap1, cv2.COLORMAP_JET)
heatmap1 = heatmap1[:, :, (2, 1, 0)]

fig = plt.figure()
# fig.add_subplot(1, 2, 1)
# # image_patches = np.asarray(255 * torch.squeeze(image_patches).cpu(), dtype='uint8').transpose((1, 2, 0))
# # plt.imshow(image_patches)
# plt.axis('off')
# plt.gca().add_patch(plt.Rectangle((x_comp - 2, y_comp - 2), 2, 2, color='red', fill=False, linewidth=1))

# fig.add_subplot(1, 2, 2)
plt.imshow(heatmap1)
plt.gca().add_patch(plt.Rectangle((x_comp - 2, y_comp - 2), 2, 2, color='red', fill=False, linewidth=1))
plt.axis('off')
plt.savefig('outs.pdf', dpi=1200)
plt.show()
```
Looking forward to your reply.
Thanks again!

**Comment by henuJackSon** [comment]
The output results are as follows:
![image](https://github.com/user-attachments/assets/2bdcc9d5-4e3b-4091-8182-347967cd0087)

**Comment by sstary** [comment]
Hi, I can't tell from your code alone why you cannot draw a heat map. However, I did not store the feature map and then load it as you did, but directly obtained the feature map of the model in the test process and drew it directly. I suggest you strictly follow the process in my code to reproduce it.

**Comment by sstary** [comment]
你需要删除这一行:with torch.no_grad

**Comment by henuJackSon** [comment]
Thank you very much for replying to my message at this time. I will take your suggestion and conduct the experiment as soon as possible.
Thanks again

**Comment by sstary** [comment]
If it helps you, give me a star please.

**Comment by henuJackSon** [comment]
Sure, it's lit up the stars.

**Comment by henuJackSon** [comment]
Hello, I have a question to ask you. I found that some pictures of the calculated heat map pay more attention to those non-target areas. In particular, the better the accuracy of the model, the more detailed the shape, edge and details of the non-target area are displayed on the heat map. However, the target area is blue. Is this the cause? The reason I am thinking of now is whether it is related to the channel selected by 'pred' and the selected coordinate point?
Looking forward to your reply
Thank you
```python
pred = out_bin[:, 1, 100, 65]
```

**Comment by sstary** [comment]
The focused area is red area，and it is determined by the second value in pred. For example, pred=[:，1，100，50] means to find the building related area since 1 refer to the building class as we defined in utils.py.

**Comment by henuJackSon** [comment]
Thank you very much for your prompt reply. Can I understand it this way? Heatmaps can only visualize 1 class at a time. According to the prediction results output by the decoder, the channel corresponding to the category is selected and visualized. One question I had was that my task could be reduced to a binary classification task, where the encoder outputs change and no change classes. Can I view the changes as a category and heat map the entire area of change?
Looking forward to your reply
Thanks again

**Comment by sstary** [comment]
yes.
